1300 filename normalization doesn't work for removes

yuripv opened the pull request:
Work by Kevin Crowe (@kcrowenex).

ahrens commented:
@zettabot go

ahrens commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -87,22 +88,17 @@ extern "C" {
+ * MT_EXACT - exact match, bypasses normalization to do strcmp()
+ * MT_FIRST - Normalizes the lookup - compose/decompose and then
+ *            match the normalized string.
+ * MT_CASE - Used to keep track of whether the lookup request is
```
if MT_CASE is set, then it is case-sensitive?

kcrowenex replied:
On Mon, Mar 28, 2016 at 6:38 PM, Matthew Ahrens <notifications@github.com>
wrote:

> In usr/src/uts/common/fs/zfs/sys/zap.h
> <https://github.com/openzfs/openzfs/pull/77#discussion_r57645678>:
>
> > @@ -87,22 +88,17 @@ extern "C" {
> >
> >  /*
> >   * Specifies matching criteria for ZAP lookups.
> > + * MT_EXACT - exact match, bypasses normalization to do strcmp()
> > + * MT_FIRST - Normalizes the lookup - compose/decompose and then
> > + *            match the normalized string.
> > + * MT_CASE - Used to keep track of whether the lookup request is
>
> if MT_CASE is set, then it is case-sensitive?
>

Correct.  Ultimately I believe you are asking/suggesting the comment might
be improved to indicate that if it's set then that means the lookup is case
sensitive?  I'd be hapy with that change as it helps the reader understand
the code better.



> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openzfs/openzfs/pull/77/files/713d9ff5d7adc49d9e5858a05121641625b6f8d3#r57645678>
>

ahrens replied:
Yes, and/or change the macro name to MT_CASE_INSENSITIVE.

kcrowenex replied:
I'm OK with a name change, but it's actually SENSITIVE, not INSENSITIVE.  Sorry for the length, but here's why:

The root of all this is obviously that only the match type of FIRST or EXACT (or the BEST I removed) was kept track of down the stack calls to the normalization and comparison code (and that is what resulted in the flip-flopping in the past of folks fixing case and breaking normalization.)  Thus I had to wedge additional info (CASE) into the match type (mt) as there are 2 cases on a normalized lookup where case matters - I'll refer you to zfs_dropname() (same logic is in zfs_dirent_lock() ) and the header comment above the zfs_dropname function.  Those cases are commented on lines 795 and 798 of zfs_dir.c.  These:

795  * CI  norm | z_norm = upper|formD | mt = MT_FIRST | MT_CASE  (ZCIEXACT)
798  * CM  norm | z_norm = upper|formD | mt = MT_FIRST | MT_CASE  (unix = !ZCILOOK)

Line 795 comments a situation that (probably?) never occurs:  a case insensitive & normalized file system for which the ZCIEXACT flag is passed in?  That implies that we have to normalize (MT_FIRST lookup) but becasue of ZCIEXACT case matters (so we or in MT_CASE.)

Line 798 is more common:  A mixed case & normalized fs.  ZCILOOK is a flag that comes from the Windows (SMB) entry into this lookup path, Unix never sets that flag.  So, if it's a lookup on this fs then case matters (so we or in MT_CASE.)

Now the key use for MT_CASE is down in zap_name_alloc(), and that was the real crux of this problem.  That U8_TEXTPREP_TOUPPER flag gets set on the fs on creation (if I remember correctly?) and it was the normalize and subsequent calls that needed to _preserve_ case on these lookups by removing U8_TEXTPREP_TOUPPER.  MT_CASE is the flag I use to indicate this.

Sorry for the length, but I wanted to make sure it was clear why MT_CASE is sensitive.  That is, a case sensitive lookup on a normalized fs.

So...you OK with changing it to MT_CASE_SENSITIVE ?

ahrens replied:
I'm certainly OK with changing the name to be more accurate (MT_CASE -> MT_CASE_SENSITIVE).  I'm still not sure I follow why the "sense" of the two bits has to be opposite.  For clarity it would be preferable if both bits indicated either "ignore this aspect" or "pay attention to this aspect".

I.e. either (MT_CASE_SENSITIVE, MT_FORM_SENSITIVE) or (MT_CASE_INSENSITIVE, MT_FORM_INSENSITIVE (or MT_NORMALIZED))

ahrens commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -87,22 +88,17 @@ extern "C" {
-	 */
-	MT_FIRST
+	MT_EXACT = 0x1,
+	MT_FIRST,
```
If this is supposed to be a bitfield, then please explicitly set this to 0x2.

kcrowenex replied:
On Mon, Mar 28, 2016 at 6:39 PM, Matthew Ahrens <notifications@github.com>
wrote:

> In usr/src/uts/common/fs/zfs/sys/zap.h
> <https://github.com/openzfs/openzfs/pull/77#discussion_r57645719>:
>
> >   */
> >  typedef enum matchtype
> >  {
> > -	/* Only find an exact match (non-normalized) */
> > -	MT_EXACT,
> > -	/*
> > -	 * If there is an exact match, find that, otherwise find the
> > -	 * first normalized match.
> > -	 */
> > -	MT_BEST,
> > -	/*
> > -	 * Find the "first" normalized (case and Unicode form) match;
> > -	 * the designated "first" match will not change as long as the
> > -	 * set of entries with this normalization doesn't change.
> > -	 */
> > -	MT_FIRST
> > +	MT_EXACT = 0x1,
> > +	MT_FIRST,
>
> If this is supposed to be a bitfield, then please explicitly set this to
> 0x2.
>

Agreed.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openzfs/openzfs/pull/77/files/713d9ff5d7adc49d9e5858a05121641625b6f8d3#r57645719>
>

ahrens commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -87,22 +88,17 @@ extern "C" {
 
 /*
  * Specifies matching criteria for ZAP lookups.
```
Is this trying to keep track of 2 things:
1. should we do UTF form normalization?
2. should we do case insensitivity (can also be thought of as "case normalization")?
If so, can we represent this with a bitfield with 2 entries:
1. MT_NORMALIZED_FORM
2. MT_CASE_INSENSITIVE (or MT_NORMALIZED_CASE)
If neither bit is set, then we are looking for an exact (strcmp()) match.

kcrowenex replied:
On Mon, Mar 28, 2016 at 6:44 PM, Matthew Ahrens <notifications@github.com>
wrote:

> In usr/src/uts/common/fs/zfs/sys/zap.h
> <https://github.com/openzfs/openzfs/pull/77#discussion_r57646377>:
>
> > @@ -87,22 +88,17 @@ extern "C" {
> >
> >  /*
> >   * Specifies matching criteria for ZAP lookups.
>
> Is this trying to keep track of 2 things:
> 1. should we do UTF form normalization?
> 2. should we do case insensitivity (can also be thought of as "case
> normalization")?
> If so, can we represent this with a bitfield with 2 entries:
> 1. MT_NORMALIZED_FORM
> 2. MT_CASE_INSENSITIVE (or MT_NORMALIZED_CASE)
> If neither bit is set, then we are looking for an exact (strcmp()) match.
>

Yes, and probably.  Yuri, if you are willing to make that change I fully
agree with Matt that it looks like it should work.



> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openzfs/openzfs/pull/77/files/713d9ff5d7adc49d9e5858a05121641625b6f8d3#r57646377>
>

ahrens commented on usr/src/uts/common/fs/zfs/zap_micro.c:
```diff
@@ -201,6 +218,19 @@ zap_name_alloc(zap_t *zap, const char *key, matchtype_t mt)
+		 * We *must* use zn_normflags because this normalization is
+		 * what the matching is based on.  (Not the hash!)
+		 */
+		if (zap_normalize(zap, key, zn->zn_normbuf,
```
Don't you need to change zn_key_norm_numints now, since zn_key_norm (==zn_normbuf) could be a different length now?

kcrowenex replied:
On Mon, Mar 28, 2016 at 6:54 PM, Matthew Ahrens <notifications@github.com>
wrote:

> In usr/src/uts/common/fs/zfs/zap_micro.c
> <https://github.com/openzfs/openzfs/pull/77#discussion_r57647525>:
>
> > @@ -201,6 +218,19 @@ zap_name_alloc(zap_t *zap, const char *key, matchtype_t mt)
> >  	}
> >
> >  	zn->zn_hash = zap_hash(zn);
> > +
> > +	if (zap->zap_normflags) {
> > +		/*
> > +		 * We *must* use zn_normflags because this normalization is
> > +		 * what the matching is based on.  (Not the hash!)
> > +		 */
> > +		if (zap_normalize(zap, key, zn->zn_normbuf,
>
> Don't you need to change zn_key_norm_numints now, since zn_key_norm
> (==zn_normbuf) could be a different length now?
>

Ewww....yeah, good catch.  I believe you are correct.



> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openzfs/openzfs/pull/77/files/713d9ff5d7adc49d9e5858a05121641625b6f8d3#r57647525>
>

yuripv replied:
I don't think that it's an issue, the only possible difference between zap->zap_normflags and zn->zn_normflags is U8_TEXTPREP_TOUPPER flag, which shouldn't affect the resulting length.

yuripv replied:
Fixed anyway.

ahrens commented:
@kcrowenex @yuripv any interest in making the changes discussed here?

kcrowenex commented:
Yes, but...

Right now I'm drowning in work and my allegiance is to the company that
helps
me to pay my mortgage.  I can't justify even a few minutes on work that
doesn't go towards our current goal.  Meanwhile, the fix we have in-house
for
this works so it's hard for me to justify changing it for our own benefit.

My free time outside of work doesn't involve illumos or openzfs as I have
other interests.  The few times I thought "I'll just make that change Sunday
afternoon and test it..."  Well, Sunday afternoon comes and goes and I have
zero desire to touch that code.

Do we have other options?
-Wait another month and see if I get a few moments at my day job for this
 rework?
-Do you (or someone else?) want to make the changes per the suggestions
 you made?  I've no problem with someone else making the MT_* changes
 you suggested.
-Other ideas?

Thanks,
Kevin.



On Sat, May 14, 2016 at 11:33 PM, Matthew Ahrens <notifications@github.com>
wrote:

> @kcrowenex <https://github.com/kcrowenex> @yuripv
> <https://github.com/yuripv> any interest in making the changes discussed
> here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openzfs/openzfs/pull/77#issuecomment-219264297>
>

ahrens commented:
@kcrowenex Thanks for the update, I understand the lack of time.  I probably won't have time to get to this soon either.  It would be great if someone who uses case normalization at $work could pick this up and help get it integrated.

yuripv commented:
Just a rebase for now, will try addressing @ahrens review comments.

yuripv commented:
Addressed comments from @ahrens (I hope):
- size of the normbuf is recalculated.
- removed MT_EXACT, leaving only MT_CASE and MT_FORM describing case sensitivity and normalization - didn't use the long proposed names for the sake of readability.
- changed the matchtype_t into an int as or'ing the flags is out of bound of declared enum.

ahrens commented:
@zettabot go

yuripv commented:
Somewhat unrelated question.. Looking at the zfs-tests results - does anyone know when zfs_acl_chmod_xattr_001_pos started failing? Wonder if it's test case problem or some kind of regression.

jwk404 commented:
@yuripv I haven't seen that test fail prior to the first week in November.

ahrens commented on usr/src/uts/common/fs/zfs/zap_leaf.c:
```diff
@@ -697,7 +688,7 @@ zap_entry_normalization_conflict(zap_entry_handle_t *zeh, zap_name_t *zn,
 
 		if (zn == NULL) {
-			zn = zap_name_alloc(zap, name, MT_FIRST);
+			zn = zap_name_alloc(zap, name, MT_FORM);
```
I'm not clear on why this is MT_FORM vs MT_CASE vs based on some defaults in the ZAP.  It looks like this is used (with `zn==NULL`) only by `fzap_cursor_retrieve()`.

ahrens commented on usr/src/uts/common/fs/zfs/zap_micro.c:
```diff
@@ -156,15 +157,15 @@ zap_match(zap_name_t *zn, const char *matchname)
 	ASSERT(!(zap_getflags(zn->zn_zap) & ZAP_FLAG_UINT64_KEY));
 
-	if (zn->zn_matchtype == MT_FIRST) {
+	if (zn->zn_matchtype & MT_FORM) {
```
We don't need to normalize if we're case-insensitive (i.e. `MT_CASE` is set)?

yuripv replied:
No, same as previously, this isn't about being case-insensitive.

ahrens commented on usr/src/uts/common/fs/zfs/zfs_dir.c:
```diff
@@ -175,13 +174,27 @@ zfs_dirent_lock(zfs_dirlock_t **dlpp, znode_t *dzp, char *name, znode_t **zpp,
+	 * When matching we may need to normalize & change case and in addition
+	 * we need to honor the lookup request during the match.  First
+	 * determine if we need to normalize this text by looking for a FIRST
+	 * match, or if this is an EXACT match, bypassing normalization.  Then
```
Given that `MT_FIRST` and `MT_EXACT` no longer exist, does `FIRST` and `EXACT` refer to something specific here?

yuripv replied:
Will update the comment.

yuripv commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -88,23 +89,15 @@ extern "C" {
+enum {
+	MT_CASE = 0x1,
+	MT_FORM = 0x2
+};
```
Ok.

yuripv commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -208,20 +191,18 @@ int zap_lookup(objset_t *ds, uint64_t zapobj, const char *name,
-    matchtype_t mt, char *realname, int rn_len,
-    boolean_t *normalization_conflictp);
+    uint64_t integer_size, uint64_t num_integers, void *buf, int mt,
+    char *realname, int rn_len, boolean_t *normalization_conflictp);
```
Ok, will revert.

yuripv commented on usr/src/uts/common/fs/zfs/zap_leaf.c:
```diff
@@ -361,7 +362,7 @@ zap_leaf_array_match(zap_leaf_t *l, zap_name_t *zn,
 
 	ASSERT(zn->zn_key_intlen == 1);
-	if (zn->zn_matchtype == MT_FIRST) {
+	if (zn->zn_matchtype & MT_FORM) {
```
MT_CASE isn't about ignoring the case, it's about honoring (preserving) the case in lookup request, so that we reply with what we were asked about if there's a match.

All normalization settings (both unicode and case-folding) are specified in other flags, MT_FORM specifies whether we should use the flags specified in z_norm, MT_CASE is about honoring the lookup case.

yuripv commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -88,23 +89,15 @@ extern "C" {
 /*
  * Specifies matching criteria for ZAP lookups.
+ * MT_CASE - Used to keep track of whether the lookup request is
+ *           case sensitive or not (i.e. Windows vs Unix.)
```
Same as below.

First flag is about using the FS normalization settings (both unicode and case), so MT_IGNORE_FORM makes sense here, will rename.

Second flag is about honoring the case in lookup request, so it should be something like MT_PRESERVE_CASE.

yuripv commented on usr/src/uts/common/fs/zfs/zap_leaf.c:
```diff
@@ -418,9 +418,9 @@ zap_leaf_lookup(zap_leaf_t *l, zap_name_t *zn, zap_entry_handle_t *zeh)
 		 * NB: the entry chain is always sorted by cd on
 		 * normalized zap objects, so this will find the
-		 * lowest-cd match for MT_FIRST.
+		 * lowest-cd match for MT_FORM.
```
No, hopefully my description of the flags answers why.

ahrens commented:
@zettabot go

ahrens commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -88,22 +89,12 @@ extern "C" {
 
 /*
  * Specifies matching criteria for ZAP lookups.
+ * MT_IGNORE_FORM - normalize the lookup according to FS settings.
```
I don't think the ZAP is examining the FS settings, it's looking at the ZAP settings, specifically zap_normflags, which can be `U8_TEXTPREP_NF* | U8_TEXTPREP_TOUPPER`.  Could you expand the comment to include that?  In particular, that this can include both unicode normalization and case-insensitivity.  I wonder if this would be better named `MT_NORMALIZE`.

yuripv replied:
Done.

ahrens commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -88,22 +89,12 @@ extern "C" {
 /*
  * Specifies matching criteria for ZAP lookups.
+ * MT_IGNORE_FORM - normalize the lookup according to FS settings.
+ * MT_PRESERVE_CASE - honor (preserve) the case in the lookup.
```
I think that this flag is only used when MT_IGNORE_FORM is also specified, and it has the effect of doing a case-sensitive lookup even though MT_IGNORE_FORM is specified and zap_normflags includes U8_TEXTPREP_TOUPPER.  If that's right, can you expand the comment to explain that?

I think this would be better named `MT_MATCH_CASE` or `MT_CASE_SENSITIVE`.  "Preserve case" means something else: when we create an entry, we "preserve case", i.e. store it in the case specified, even though lookups can be case-insensitive.  This way, directory listings preserve the case that was used when the entry was created.

yuripv replied:
OK.

ahrens commented on usr/src/uts/common/fs/zfs/sys/zap.h:
```diff
@@ -120,16 +111,6 @@ typedef enum zap_flags {
- * U8_TEXTPREP_TOLOWER: case normalization will be performed.
- *     MT_FIRST/MT_BEST matching will find entries that match without
- *     regard to case (eg. looking for "foo" can find an entry "Foo").
- * Eventually, other flags will permit unicode normalization as well.
```
I agree that this comment is wrong and should be deleted.  Is there somewhere else that we document the `normflags`?  If not, could you add it to `zap_create_norm` in `zap_micro.c` or to `mz_normflags` / `zap_normflags` in zap_impl.h.?  I think the documentation should list what flags are acceptable, specifically one of `U8_TEXTPREP_NF*` plus `U8_TEXTPREP_TOUPPER`

yuripv replied:
Could you please help me here a bit with a correct wording (as I'm really not a ZFS guy, for now, at least)?

ahrens replied:
How about adding this comment above mzap_create_impl() in zap_micro.c:

```
/*
 * The "normflags" determine the behavior of the matchtype_t (which is
 * passed to zap_lookup_norm()).  Names which have the same normalized
 * version will be stored with the same hash value, and therefore we can
 * perform normalization-insensitive lookups.  We can be Unicode form-
 * insensitive and/or case-insensitive.  The following flags are valid for
 * "normflags":
 * 
 * U8_TEXTPREP_NFD
 * U8_TEXTPREP_NFC
 * U8_TEXTPREP_NFKD
 * U8_TEXTPREP_NFKC
 * U8_TEXTPREP_TOUPPER
 * 
 * The *_NF* (Normalization Form) flags are mutually exclusive; at most one
 * of them may be supplied.
 */
```

ahrens commented on usr/src/uts/common/fs/zfs/zap_micro.c:
```diff
@@ -202,6 +218,20 @@ zap_name_alloc(zap_t *zap, const char *key, matchtype_t mt)
 
 	zn->zn_hash = zap_hash(zn);
+
+	if (zap->zap_normflags) {
```
I wonder if it would be worth skipping this if `zn_normflags == zap_normflags`(i.e. if they didn't specify `MT_PRESERVE_CASE`) (for performance reasons)?

yuripv replied:
Done.

ahrens commented on usr/src/uts/common/fs/zfs/zfs_dir.c:
```diff
@@ -774,25 +789,49 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
+ * fs type  | z_norm      | lookup type | match type
+ * ---------|-------------|-------------|----------------------------------
+ * CS !norm | 0           |           0 | exact
+ * CS  norm | formD       |           0 | MT_IGNORE_FORM
```
Why is this `formD`, as opposed to using the `normalization` property (indicated here by `formX`)?

yuripv replied:
Fixed.

ahrens commented on usr/src/uts/common/fs/zfs/zfs_dir.c:
```diff
@@ -774,25 +789,49 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
+ * ------------------------------------------------------------------------
+ * fs type  | z_norm      | lookup type | match type
+ * ---------|-------------|-------------|----------------------------------
+ * CS !norm | 0           |           0 | exact
```
Maybe this match type should be listed as `0 (exact)`

yuripv replied:
Done.

ahrens commented on usr/src/uts/common/fs/zfs/zfs_dir.c:
```diff
@@ -774,25 +789,49 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
+ * CM !norm | upper       |    !ZCILOOK | MT_IGNORE_FORM | MT_PRESERVE_CASE
+ * CM !norm | upper       |     ZCILOOK | MT_IGNORE_FORM
+ * CM  norm | upper|formX |    !ZCILOOK | MT_IGNORE_FORM | MT_PRESERVE_CASE
+ * CM  norm | upper|formX |     ZCILOOK | MT_IGNORE_FORM
```
Interface-wise, I think it would be easier to understand the interface I misunderstood you to be implementing originally: two flags each of which specifies one orthogonal aspect to be normalized/ignored:
`MT_NORMALIZE_UNICODE_FORM` and `MT_CASE_INSENSITIVE` (or perhaps named `MT_IGNORE_FORM` and `MT_IGNORE_CASE`)

The table would look the same as this except the `match type` would substitute: `MT_IGNORE_FORM | MT_PRESERVE_CASE` -> `MT_NORMALIZE_UNICODE_FORM`
`MT_IGNORE_FORM` -> `MT_NORMALIZE_UNICODE_FORM | MT_CASE_INSENSITIVE`

I think what you've done here is still a good improvement on the existing logic so I won't insist on this additional change, but I think it would further improve understandability of the code.

yuripv commented:
Added requested changes except for documenting zap normalization flags.

ahrens commented:
@zettabot go

yuripv commented:
Thanks Matt, comment added.

yuripv commented:
Can we build this again?

prakashsurya commented:
@zettabot go

yuripv commented:
I've pushed hopefully the final  update here, the only functional change (among the copyrights update, as the work somewhat differs from what it was originally) is the check in zap_name_alloc():

if (zap->zap_normflags != zn->zn_normflags)

instead of

if (zap->zap->normflags && zap->zap_normflags != zn->zn_normflags).

Please rerun the build and sorry for the late push.

prakashsurya commented:
@zettabot go

yuripv commented:
Please re-run.

ahrens commented:
@zettabot go

yuripv commented:
Anything else I need to do here?

ahrens commented:
@yuripv Not waiting on anything from you.  I am trying to get one more reviewer, I'm asking around at Delphix.  Look for another review in the coming week.

pzakha commented on usr/src/uts/common/fs/zfs/zap_micro.c:
```diff
@@ -185,15 +187,30 @@ zap_name_alloc(zap_t *zap, const char *key, matchtype_t mt)
 	zn->zn_key_orig = key;
 	zn->zn_key_orig_numints = strlen(zn->zn_key_orig) + 1;
 	zn->zn_matchtype = mt;
+	zn->zn_normflags = zap->zap_normflags;
```
I think we could move this after `zn->zn_hash = zap_hash(zn);` for readability purposes.

yuripv commented:
Excerpt from zfs-tests log, these are the ones that were expected to fail without this fix

```
    PASS casenorm/insensitive_formd_delete (expected FAIL)
    PASS casenorm/sensitive_formd_delete (expected FAIL)
    PASS casenorm/insensitive_none_delete (expected FAIL)
    PASS casenorm/mixed_formd_lookup (expected FAIL)
    PASS casenorm/mixed_formd_delete (expected FAIL)
```
